python/gs/ops/sddmm.py

```diff
@@ -0,0 +1,180 @@
+
+    .. math::
+
+        x_{e} = \phi(x_{lhs}, x_{rhs}), \forall (u,e,v)\in \mathcal{G}
+
+    where :math:`x_{e}` is the returned feature on edges and :math:`x_u`,
+    :math:`x_v` refers to :attr:`u`, :attr:`v` respectively. :math:`\phi`
+    is the binary operator :attr:`op`, and :math:`\mathcal{G}` is the graph
+    we apply gsddmm on: :attr:`g`. :math:`lhs` and :math:`rhs` are one of
+    :math:`u,v,e`'s.
+
+    Parameters
+    ----------
+    g : gs.Matrix
+        The input graph.
+    op : str
+        Binary operator, could be ``add``, ``sub``, ``mul``, ``div``, ``dot``,
+        ``copy_lhs``, ``copy_rhs``.
+    lhs_data : tensor or None
+        The left operand, could be None if it's not required by op.
+    rhs_data : tensor or None
+        The right operand, could be None if it's not required by op.
+    lhs_target: str
+        Choice of ``u``(source), ``e``(edge) or ``v``(destination) for left operand.
+    rhs_target: str
+        Choice of ``u``(source), ``e``(edge) or ``v``(destination) for right operand.
+
+    Returns
+    -------
+    tensor
+        The result tensor.
+    """
+    if op not in ['copy_lhs', 'copy_rhs']:
+        lhs_data, rhs_data = reshape_lhs_rhs(lhs_data, rhs_data)
+    return gsddmm_internal(
+        g._graph, op, lhs_data, rhs_data, lhs_target, rhs_target)
+
+
+def _gen_sddmm_func(lhs_target, rhs_target, binary_op):
+    name = "{}_{}_{}".format(lhs_target, binary_op, rhs_target)
+    target_dict = {
+        'u': "source node",
+        'e': "edge",
+        'v': "destination node"
+    }
+    lhs_str = target_dict[lhs_target]
+    rhs_str = target_dict[rhs_target]
+    docstring = r"""Generalized SDDMM function.
+    It computes edge features by {op} {lhs} features and {rhs} features.
+
+    Parameters
+    ----------
+    g : DGLHeteroGraph
+        The input graph
+    x : tensor
+        The {lhs} features.
+    y : tensor
+        The {rhs} features.
+
+    Returns
+    -------
+    tensor
+        The result tensor.
+
+    Notes
+    -----
+    If the feature shape of two input operands do not match, we first broadcasts the features to a unified
+    shape (note that the memory usage will not increase accordingly) and then performs the operation.
+
+    Broadcasting follows NumPy semantics. Please see
+    https://docs.scipy.org/doc/numpy/user/basics.broadcasting.html
+    for more details about the NumPy broadcasting semantics.
+    """.format(op=binary_op, lhs=lhs_str, rhs=rhs_str)
+
+    def func(g, x, y):
+        return gsddmm(g, binary_op, x, y,
+                      lhs_target=lhs_target, rhs_target=rhs_target)
+    func.__name__ = name
+    func.__doc__ = docstring
+    return func
+
+
+def _register_sddmm_func():
+    """Register sddmm functions"""
+    target = ["u", "v", "e"]
+    for lhs, rhs in product(target, target):
+        if lhs != rhs:
+            for binary_op in ["add", "sub", "mul", "div", "dot"]:
```

Comment:
Do we need to handle all these cases? Supporting `dot` may be enough.

Reply:
In AS-GCN, both `add` and `div` is used.